### ISSUE TITLE
Shrink ticket-rag image: install CPU-only torch + audit other heavy deps

### ISSUE THREAD

**Issue by iansmith** [iansmith/ticket-plugin#21]
# BILL-21 — Shrink ticket-rag image: install CPU-only torch + audit other heavy deps

**Ticket system:** GitHub
**State:** OPEN, In Progress (status:in-progress)
**Labels:** enhancement, status:in-progress
**Parent / umbrella:** [#13](https://github.com/iansmith/ticket-plugin/issues/13)
**Sibling subtasks:** #12 (done), #14 (done), #16 (done), #15, #17, #18
**Ticket URL:** https://github.com/iansmith/ticket-plugin/issues/21
**Started:** 2026-05-26

## Original description (snapshot at start)

**Subtask of:** #13 (umbrella — ticket-rag service container).

## Summary

The image built by [BILL-14](https://github.com/iansmith/ticket-plugin/issues/14) lands at ~5.77 GB before models are even baked in, well above the design doc's "~3 GB with models" target. The dominant cost is `torch` pulling NVIDIA CUDA wheels (`nvidia-cublas-cu13`, `nvidia-cudnn-cu13`, etc., ~3 GB total) as transitive deps of `sentence-transformers`. These wheels are useless for our use case: localhost-only RAG sidecar, CPU-only encoder + reranker, no GPU in the planned deployment.

Switch to PyTorch's CPU-only wheel index and audit any other heavy transitive dependencies.

## Why now

- Slow `docker build` iterations during BILL-15 / BILL-17 development.
- Significant local disk consumption per build.
- If we ever publish the image to a registry (out of scope today, but possible), bandwidth + storage cost scales with image size.
- The design doc explicitly targets ~3 GB; we're already 2× over before models.

### 1. CPU-only torch

Edit `docker/postgres-pgvector/requirements.txt` to pin torch from the CPU-only wheel index:

```
--extra-index-url https://download.pytorch.org/whl/cpu
torch==2.5.1+cpu
sentence-transformers==3.3.1
```

Exact syntax may need adjustment for pip's resolver — verify the chosen torch version actually resolves against sentence-transformers' constraints and against linux/arm64 + linux/amd64.

### 2. Audit other heavy deps

Run `pip install -r requirements.txt && pip show <pkg>` for each dep, plus `du -sh /usr/local/lib/python3.12/site-packages/*` inside the container, to identify any remaining > 100 MB transitive deps that might be optional or replaceable.

### 3. Optional: `--no-install-recommends` audit on apt installs

BILL-14 added `libgomp1` via `apt-get install -y --no-install-recommends`. Confirm this is the only apt addition and that no `Recommends:`-style transitive packages crept in.

## Acceptance (carried from ticket)

- [ ] Image size after `docker build docker/postgres-pgvector/` is documented and reported; target: under 3 GB before BILL-15 model bake-in.
- [ ] `pip freeze` inside the container shows `torch` from the `+cpu` index (or equivalent — verify via `pip show torch` showing the right wheel URL).
- [ ] Both BILL-14 acceptance criteria still pass: `/healthz` happy path returns 200, degraded path returns 503.
- [ ] Smoke test: import `sentence_transformers` inside the container with no error.
- [ ] requirements.txt header comment notes the CPU-only constraint and why (so future contributors don't "fix" it by removing the index URL).

## Out of scope (carried from ticket)

- Splitting the service container into separate postgres + python images.
- Switching encoder/reranker models.
- Stripping models from the image and downloading them at runtime.

## Plan

Delegated to a worktree-isolated agent per the orchestration pattern. Phases:

1. **Phase 0** — read the BILL-14 Dockerfile and requirements.txt; measure the current image size as a baseline.
2. **Phase 1** — edit `requirements.txt` to use the PyTorch CPU-only wheel index. Verify pip resolution against `sentence-transformers==3.3.1`.
3. **Phase 2** — rebuild and measure new image size. Verify `pip show torch` reports the +cpu wheel.
4. **Phase 3** — audit `du -sh /usr/local/lib/python3.12/site-packages/*` for remaining > 100 MB packages. Surface findings without acting (any further optimization is a follow-on ticket unless it's a one-liner).
5. **Phase 4** — re-run BILL-14's acceptance checks: `/healthz` 200, sentence_transformers loads without error. Capture image-size delta.
6. **Phase 5** — commit on `feat/BILL-21` with the size delta in the body.

Orchestrator runs `/ticket-update` → `/ticket-pr --no-poll` → `/ticket-merge` → `/ticket-archive` after agent reports.

---

## Original description (preserved)

**Subtask of:** #13 (umbrella — ticket-rag service container).

## Summary

The image built by [BILL-14](https://github.com/iansmith/ticket-plugin/issues/14) lands at ~5.77 GB before models are even baked in, well above the design doc's "~3 GB with models" target. The dominant cost is `torch` pulling NVIDIA CUDA wheels (`nvidia-cublas-cu13`, `nvidia-cudnn-cu13`, etc., ~3 GB total) as transitive deps of `sentence-transformers`. These wheels are useless for our use case: localhost-only RAG sidecar, CPU-only encoder + reranker, no GPU in the planned deployment.

Switch to PyTorch's CPU-only wheel index and audit any other heavy transitive dependencies.

## Why now

- Slow `docker build` iterations during BILL-15 / BILL-17 development.
- Significant local disk consumption per build.
- If we ever publish the image to a registry (out of scope today, but possible), bandwidth + storage cost scales with image size.
- The design doc explicitly targets ~3 GB; we're already 2× over before models.

### 1. CPU-only torch

Edit `docker/postgres-pgvector/requirements.txt` to pin torch from the CPU-only wheel index:

```
--extra-index-url https://download.pytorch.org/whl/cpu
torch==2.5.1+cpu
sentence-transformers==3.3.1
```

Exact syntax may need adjustment for pip's resolver — verify the chosen torch version actually resolves against sentence-transformers' constraints and against linux/arm64 + linux/amd64.

### 2. Audit other heavy deps

Run `pip install -r requirements.txt && pip show <pkg>` for each dep, plus `du -sh /usr/local/lib/python3.12/site-packages/*` inside the container, to identify any remaining > 100 MB transitive deps that might be optional or replaceable. Candidates likely to surface:

- `scipy` (sentence-transformers pulls it through scikit-learn)
- `numpy` (necessary, but check if there's an arm64-optimized wheel)
- `huggingface-hub` cache directories baked into the image

Document anything removed or pinned-down in the requirements.txt header comment.

### 3. Optional: `--no-install-recommends` audit on apt installs

BILL-14 added `libgomp1` via `apt-get install -y --no-install-recommends`. Confirm this is the only apt addition and that no `Recommends:`-style transitive packages crept in.

## Acceptance

- [ ] Image size after `docker build docker/postgres-pgvector/` is documented and reported; target: under 3 GB before BILL-15 model bake-in.
- [ ] `pip freeze` inside the container shows `torch` from the `+cpu` index (or equivalent — verify via `pip show torch` showing the right wheel URL).
- [ ] Both BILL-14 acceptance criteria still pass: `/healthz` happy path returns 200, degraded path returns 503.
- [ ] Smoke test: import `sentence_transformers` inside the container with no error.
- [ ] requirements.txt header comment notes the CPU-only constraint and why (so future contributors don't "fix" it by removing the index URL).

## Out of scope

- Splitting the service container into separate postgres + python images. The umbrella ticket BILL-13 specifies a single image; multi-container architecture is a different design conversation.
- Switching encoder/reranker models. BILL-15 owns model selection.
- Stripping models from the image and downloading them at runtime. That's a deliberate design choice in the doc (bake at build time, fully offline) and should not be reversed here.

## Why this is a separate ticket from BILL-18

[BILL-18](https://github.com/iansmith/ticket-plugin/issues/18) (build pipeline) owns reproducibility, layer-cache strategy, and Taskfile orchestration. This ticket owns aggressive image-size reduction, which is a content-of-the-image concern. The two can land in either order — BILL-18's layer-cache strategy still works after the torch swap, and this ticket's audit still works after BILL-18 introduces a Taskfile.
